scylla/src/transport/topology.rs

```diff
@@ -902,6 +903,91 @@ async fn query_user_defined_types(
+            PreCqlType::Native(_) => (),
+            PreCqlType::Collection { type_, .. } => match type_ {
+                PreCollectionType::List(t) | PreCollectionType::Set(t) => {
+                    do_with_referenced_udts(what, t)
+                }
+                PreCollectionType::Map(t1, t2) => {
+                    do_with_referenced_udts(what, t1);
+                    do_with_referenced_udts(what, t2);
+                }
+            },
+            PreCqlType::Tuple(types) => types
+                .iter()
+                .for_each(|type_| do_with_referenced_udts(what, type_)),
+            PreCqlType::UserDefinedType { name, .. } => what(name),
+        }
+    }
+
+    // Build an indegree map: for each node in the graph, how many directly depending types it has.
+    let mut indeg = udts
+        .drain(..)
+        .map(|def| {
+            (
+                (def.keyspace_name.clone(), def.type_name.clone()),
+                (def, Cell::new(0u32)),
+            )
+        })
+        .collect::<HashMap<_, _>>();
+
+    // For each node in the graph...
+    for (def, _) in indeg.values() {
+        let mut increment_referred_udts = |type_name: &str| {
+            let deg_cell = indeg
+                .get(&(def.keyspace_name.clone(), type_name.to_string()))
+                .map(|(_, count)| count)
+                .unwrap();
+            deg_cell.set(deg_cell.get() + 1);
+        };
+        // For each type referred by the node...
+        for field_type in def.field_types.iter() {
+            do_with_referenced_udts(&mut increment_referred_udts, field_type);
+        }
+    }
+
+    let mut sorted = Vec::with_capacity(indeg.len());
+    let mut next_idx = 0;
+
+    // Schedule keys that had an initial indeg of 0
+    for (key, _) in indeg.iter().filter(|(_, (_, deg))| deg.get() == 0) {
+        sorted.push(key);
+    }
+
+    while let Some(key @ (keyspace, _type_name)) = sorted.get(next_idx).copied() {
+        next_idx += 1;
+        // Decrement the counters of all UDTs that this UDT depends upon
+        // and then schedule them if their counter drops to 0
+        let mut decrement_referred_udts = |type_name: &str| {
+            let ref_key = indeg
+                .get_key_value(&(keyspace.clone(), type_name.to_string()))
+                .unwrap()
+                .0;
+            let cnt = &indeg.get(ref_key).unwrap().1;
+            let new_cnt = cnt.get() - 1;
+            cnt.set(new_cnt);
+            if new_cnt == 0 {
+                sorted.push(ref_key);
+            }
+        };
+        let def = &indeg.get(key).unwrap().0;
+        for field_type in def.field_types.iter() {
+            do_with_referenced_udts(&mut decrement_referred_udts, field_type);
+        }
+    }
+
+    let owned_sorted = sorted.into_iter().cloned().collect::<Vec<_>>();
+    assert!(udts.is_empty());
+    for key in owned_sorted.into_iter().rev() {
+        udts.push(indeg.remove(&key).unwrap().0);
+    }
+
+    Ok(())
```

Comment:
Do we detect reference cycles after all?

Reply:
Good catch, it got lost somewhere in the middle of a refactor.